src/components/main-layout/ContentPanelTimer.tsx

```diff
@@ -0,0 +1,89 @@
+import React, { useContext, useEffect, useRef, useState } from "react";
+
+import styles from "./layout.module.css";
+import { SimulariumContext } from "../../simulation/context";
+import { PageContent, Module } from "../../types";
+
+import classNames from "classnames";
+import { isEqual } from "lodash";
+import ContentPanel from "./ContentPanel";
+
+export interface ContentPanelProps {
+    pageContent: PageContent;
+    currentModule: Module;
+}
+
+// these three states are used to avoid flashing the content before the
+// opacity transition starts
+enum RenderState {
+    NoRender = 1,
+    RenderFadeIn = 2,
+    FullyVisible = 3,
+}
+
+const ContentPanelTimer: React.FC<ContentPanelProps> = ({
+    pageContent,
+    currentModule,
+}) => {
+    const [renderState, setRenderState] = useState(RenderState.NoRender);
+    const previousContentRef = useRef(pageContent);
+    const contentJustChanged = !isEqual(
+        previousContentRef.current.content,
+        pageContent.content
+    );
+
+    useEffect(() => {
+        // There are pages that have the same written content, so flashing the
+        // panel is more distracting than helpful. However, since this Ref is defaulted
+        // to the first page content, this check will always be true on the first page, and
+        // will keep the content from being rendered, so it needs to be excluded
+        if (!contentJustChanged && page !== 1) {
+            previousContentRef.current = pageContent;
+            return;
+        }
+        const count: number = 1;
+        const FADE_TIME = 150;
+        let timer: NodeJS.Timeout;
+        const updateRenderState = (count: number) => {
+            previousContentRef.current = pageContent;
+            setRenderState(count);
+            count = count + 1;
+
+            if (count <= RenderState.FullyVisible) {
+                timer = setTimeout(() => updateRenderState(count), FADE_TIME);
+            } else {
+                return () => clearTimeout(timer);
+            }
+        };
+        timer = setTimeout(() => updateRenderState(count), FADE_TIME);
+        return () => clearTimeout(timer);
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [pageContent.content]);
```

Comment:
For me, declaring the constants (count, FADE_TIME) and timer inside the useEffect, but mutating them inside `updateRenderState` feels off, and the repetition from 52-59 is a bit confusing to read. I have a small refactor to suggest that pulls the variables and helper outside the `useEffect` and returns the timer directly, we shouldn't need to clean up the timer inside `updateRenderState` once we hit the base case.
```suggestion
   
    const FADE_TIME = 150;
    const INITIAL_RENDER_STATE = 1;

    const updateRenderState = (currentState: number): NodeJS.Timeout | null => {
        previousContentRef.current = pageContent;
        setRenderState(currentState);

        const nextState = currentState + 1;
        if (nextState <= RenderState.FullyVisible) {
            return setTimeout(() => updateRenderState(nextState), FADE_TIME);
        }
        return null;
    };

    useEffect(() => {
        // There are pages that have the same written content, so flashing the
        // panel is more distracting than helpful. However, since this Ref is defaulted
        // to the first page content, this check will always be true on the first page, and
        // will keep the content from being rendered, so it needs to be excluded
        if (!contentJustChanged && page !== 1) {
            previousContentRef.current = pageContent;
            return;
        }
        const timer = setTimeout(
            () => updateRenderState(INITIAL_RENDER_STATE),
            FADE_TIME
        );
        return () => clearTimeout(timer);
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [pageContent.content]);
```